### ISSUE TITLE
Subset function for data.frames, vectors and matrices

### ISSUE THREAD

**Issue by Planeshifter** [compute-io/todo#145]
In R, there is a [subset](http://www.inside-r.org/r-doc/base/subset) function to filter rows or elements from a vector, matrix or data.frame. In JS, we natively have the `.filter()` method for Arrays, but currently have no analogous functionality for data.frames or matrices. Similar to the array method, a case can be made that these functions should be attached to the prototype.

**Comment by kgryte** [comment]
Yes for a `filter`-esque `function`. In `MATLAB`, this is `find`. We have a [`compute-find`](https://github.com/compute-io/find) module, but it needs generalization. 

And no. Nothing on the `prototype`. Most of what we do is biased toward a `functional` approach. Data structures should be as minimal as possible.

For [`matrices`](https://github.com/dstructs/matrix), the only methods we support are for element access, and this is largely to compensate for a lack of operator overloading and language deficiencies compared to, say, Python or MATLAB. And further, these methods only know the concept of indices, not predicates. 

As soon as higher order object-oriented methods are available, people will use them and, while nice for lock-in, code is now dependent on a particular data structure and API. For instance, now the concept of "matrix-like" is no longer has `length`, `strides`, `shape`, and `offset` properties, but must have `x`, `y`, and `z` methods. 

We should consider `Array` methods and all other built-ins as relics, not as examples upon which to base designs. The OO approach allowed fewer variables to "pollute" the global namespace, which was important in the context of the browser where 3rd party code may or may not introduce conflicting variable names or alter behavior such that the expectations of other 3rd party code were violated. Accordingly, a smaller namespace meant a smaller surface area.

The above does not apply with the new node paradigm where all variables are scoped to their context. And even in an interactive environment, we choose the global namespace. 

Lastly, by having `filter` / `subset` / `find` as an independent module, it can be versioned independently of the data structure. 

All this stated, for something like a `data frame`, we will need to evaluate the best approach. I am slightly more open to OO there, if only because the data structure is so specialized.
